integration-tests/docker/env.go

```diff
@@ -0,0 +1,173 @@
+package docker
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/google/uuid"
+	"github.com/rs/zerolog/log"
+	"github.com/smartcontractkit/chainlink-testing-framework/logwatch"
+	tc "github.com/testcontainers/testcontainers-go"
+	"golang.org/x/sync/errgroup"
+	"sync"
+	"text/template"
+)
+
+/* common stuff that should be moved to chainlink-env or CTF */
+
+var (
+	ErrContainerSetup  = errors.New("failed to deploy the container")
+	ErrParsingTemplate = errors.New("failed to parse template")
+)
+
+// Component is a collection one or more docker containers
+// that are working as a standalone product, for example Chainlink node
+type Component interface {
+	Prefix() string
+	Containers() []tc.Container
+	Start(dockerNet string, cfg any) (Component, error)
+	Stop() error
+}
+
+// ComponentSetupFunc is a func to set up a Component for a particular docker network
+// that may contain one or more docker containers
+type ComponentSetupFunc func(dockerNet string) (Component, error)
+
+// Environment contains several Components and can set up them sequentially or in parallel
+type Environment struct {
+	network    *tc.DockerNetwork
+	lw         *logwatch.LogWatch
+	setupFuncs []ComponentSetupFunc
+	deployMu   *sync.Mutex
+	Components []Component
+	pushToLoki bool
+}
+
+// NewEnvironment creates a new Environment
+func NewEnvironment(lw *logwatch.LogWatch, pushToLoki bool) *Environment {
```

Comment:
@skudasov this should be in CTF repo. But, I'm curious if this `Environment` can be used in the same way as the [builder struct I created for Mercury](https://github.com/smartcontractkit/mercury-server/blob/main/e2e_tests/docker/test_env/test_env_builder.go). Note how I connect each component with each other in `Build()` https://github.com/smartcontractkit/mercury-server/blob/main/e2e_tests/docker/test_env/test_env_builder.go#L94. How I can do it with `Environment` here?

Reply:
Yeah, you can't configure *all* the stuff inside, currently, but we'll work on it.
But you can add your containers and deploy them when you wish.
Ideally, we don't configure a lot of dynamic stuff and mostly use common cluster with adding stuff on top of it.
```
	env, err := docker.NewCommonChainlinkCluster(t, 1)
	require.NoError(t, err)
	env.Add(...)
	// do something else
	env.Add(...).Start(true)
```